src/Grpc.AspNetCore.Server/Internal/ServerCallDeadlineManager.cs

```diff
@@ -0,0 +1,264 @@
+            if (!TryStartExceededDeadline())
+            {
+                return;
+            }
+
+            try
+            {
+                await _serverCallContext.DeadlineExceededAsync();
+                lock (this)
+                {
+                    IsCallComplete = true;
+
+                    // Canceling CTS will trigger registered callbacks.
+                    // Exception could be thrown from them.
+                    _deadlineCts?.Cancel();
+                }
+            }
+            catch (Exception ex)
+            {
+                GrpcServerLog.DeadlineCancellationError(_serverCallContext.Logger, ex);
+            }
+            finally
+            {
+                _deadlineExceededCompleteTcs!.TrySetResult(null);
+            }
+        }
+
+        private bool TryStartExceededDeadline()
+        {
+            lock (this)
+            {
+                // Deadline callback could be raised by the CTS after call has been completed (either successfully, with error, or aborted)
+                // but before deadline exceeded registration has been disposed
+                if (!IsCallComplete)
+                {
+                    _deadlineExceededCompleteTcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    return true;
+                }
+
+                return false;
+            }
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            // Timer.DisposeAsync will wait until any in-progress callbacks are complete.
+            // By itself, this isn't enough to ensure the deadline has finished being raised because
+            // the callback doesn't return Task. _deadlineExceededCompleteTcs must also be awaited
+            // (it is set when a deadline starts being raised) to ensure the deadline manager is finished
+            // and resources can be disposed.
+
+            var disposeTask = _longDeadlineTimer.DisposeAsync();
+
+            if (disposeTask.IsCompletedSuccessfully &&
+                (_deadlineExceededCompleteTcs == null || _deadlineExceededCompleteTcs.Task.IsCompletedSuccessfully))
+            {
+                // Fast-path to avoid async state machine.
+                DisposeCore();
+                return default;
+            }
+
+            return DeadlineDisposeAsyncCore(disposeTask);
+        }
+
+        private async ValueTask DeadlineDisposeAsyncCore(ValueTask disposeTask)
+        {
+            await disposeTask;
+            // Ensure an in-progress deadline is finished before disposing.
+            // Need to await to avoid race between canceling CT and disposing it.
+            if (_deadlineExceededCompleteTcs != null)
+            {
+                await _deadlineExceededCompleteTcs.Task;
+            }
+            DisposeCore();
+        }
+
+        private void DisposeCore()
+        {
+            _deadlineCts?.Dispose();
+            _requestAbortedRegistration.Dispose();
```

Comment:
Since the callback associated with _requestAbortedRegistration Cancels _deadlineCts, this should probably be disposed before _deadlineCts.